cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -1093,52 +1057,94 @@ abstract private class SsaDefOrUse extends TSsaDefOrUse {
   abstract Location getLocation();
 }
 
-class DefOrUse extends TDefOrUse, SsaDefOrUse {
-  DefOrUseImpl defOrUse;
-
-  DefOrUse() { this = TDefOrUse(defOrUse) }
+abstract class Def extends SsaDef, TDef {
+  DefinitionExt def;
 
-  final override DefOrUseImpl asDefOrUse() { result = defOrUse }
+  Def() { this = TDef(def) }
 
-  final override Location getLocation() { result = defOrUse.getLocation() }
+  final override DefinitionExt asDef() { result = def }
 
-  final SourceVariable getSourceVariable() { result = defOrUse.getSourceVariable() }
+  /** Gets the source variable underlying this SSA definition. */
+  final SourceVariable getSourceVariable() { result = def.getSourceVariable() }
 
-  override string toString() { result = defOrUse.toString() }
+  override string toString() { result = def.toString() }
 
   /**
    * Holds if this definition (or use) has index `index` in block `block`,
    * and is a definition (or use) of the variable `sv`.
    */
   predicate hasIndexInBlock(IRBlock block, int index, SourceVariable sv) {
-    defOrUse.hasIndexInBlock(block, index, sv)
+    def.definesAt(sv, block, index, _)
   }
-}
 
-class GlobalDef extends TGlobalDef, SsaDefOrUse {
-  GlobalDefImpl global;
+  /** Gets the value written by this definition, if any. */
+  Node0Impl getValue() { none() }
 
-  GlobalDef() { this = TGlobalDef(global) }
+  /**
+   * Holds if this definition is guaranteed to overwrite the entire
+   * destination's allocation.
+   */
+  abstract predicate isCertain();
 
-  /** Gets the location of this definition. */
-  final override Location getLocation() { result = global.getLocation() }
+  /** Gets the address operand written to by this definition. */
+  Operand getAddressOperand() { none() }
 
-  /** Gets a textual representation of this definition. */
-  override string toString() { result = global.toString() }
+  /** Gets the address written to by this definition. */
+  final Instruction getAddress() { result = this.getAddressOperand().getDef() }
+
+  /** Gets the indirection index of this definition. */
+  abstract int getIndirectionIndex();
 
   /**
-   * Holds if this definition has index `index` in block `block`, and
-   * is a definition of the variable `sv`.
+   * Gets the indirection level that this definition is writing to.
+   * For instance, `x = y` is a definition of `x` at indirection level 1 and
+   * `*x = y` is a definition of `x` at indirection level 2.
+   *
+   * This predicate ensures that joins go from `defOrUse` to the result
```

Comment:
Is this text correct? `defOrUse` no longer seems a thing, right?

Reply:
Good catch. Indeed `defOrUse` is no longer a thing. I've deleted that part of the comment now. It used to have some additional `pragma`s attached to it to ensure it was join ordered correctly, but it doesn't actually need that anymore after the refactoring. So 322fa363596771e7dc22c600f115c1feac72542d just deletes that part of the comment.

Reply:
Specifically, it used to look like [this](https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll#L1247-L1250)

Reply:
Thanks